index.js

```diff
@@ -343,163 +341,159 @@ Filter.prototype._handleFile = function(relativePath, srcDir, destDir, entry, ou
 
-Filter.prototype.processFile = function(srcDir, destDir, relativePath, forceInvalidation, isChange, instrumentation, entry) {
-  let filter = this;
-  let inputEncoding = this.inputEncoding;
-  let outputEncoding = this.outputEncoding;
+    if (inputEncoding === undefined)  inputEncoding  = 'utf8';
+    if (outputEncoding === undefined) outputEncoding = 'utf8';
 
-  if (inputEncoding === undefined)  inputEncoding  = 'utf8';
-  if (outputEncoding === undefined) outputEncoding = 'utf8';
+    let contents = fs.readFileSync(srcDir + '/' + relativePath, {
+      encoding: inputEncoding
+    });
 
-  let contents = fs.readFileSync(srcDir + '/' + relativePath, {
-    encoding: inputEncoding
-  });
+    instrumentation.processString++;
+    let processStringStart = process.hrtime();
+    let string = invoke(this.processor, this.processor.processString, [this, contents, relativePath, forceInvalidation, instrumentation]);
 
-  instrumentation.processString++;
-  let processStringStart = process.hrtime();
-  let string = invoke(this.processor, this.processor.processString, [this, contents, relativePath, forceInvalidation, instrumentation]);
+    return string.then(outputString => {
+      instrumentation.processStringTime += nanosecondsSince(processStringStart);
+      let outputPath = filter.getDestFilePath(relativePath, entry);
 
-  return string.then(outputString => {
-    instrumentation.processStringTime += nanosecondsSince(processStringStart);
-    let outputPath = filter.getDestFilePath(relativePath, entry);
+      if (outputPath == null) {
+        throw new Error('canProcessFile("' + relativePath +
+                        '") is true, but getDestFilePath("' +
+                        relativePath + '") is null');
+      }
 
-    if (outputPath == null) {
-      throw new Error('canProcessFile("' + relativePath +
-                      '") is true, but getDestFilePath("' +
-                      relativePath + '") is null');
-    }
+      outputPath = destDir + '/' + outputPath;
 
-    outputPath = destDir + '/' + outputPath;
+      if (isChange) {
+        let isSame = fs.readFileSync(outputPath, 'UTF-8') === outputString;
+        if (isSame) {
+          this._logger.debug('[change:%s] but was the same, skipping', relativePath, isSame);
+          return;
+        } else {
+          this._logger.debug('[change:%s] but was NOT the same, writing new file', relativePath);
+        }
+      }
 
-    if (isChange) {
-      let isSame = fs.readFileSync(outputPath, 'UTF-8') === outputString;
-      if (isSame) {
-        this._logger.debug('[change:%s] but was the same, skipping', relativePath, isSame);
-        return;
-      } else {
-        this._logger.debug('[change:%s] but was NOT the same, writing new file', relativePath);
+      try {
+        fs.writeFileSync(outputPath, outputString, {
+          encoding: outputEncoding
+        });
+
+      } catch(e) {
+        // optimistically assume the DIR was patched correctly
```

Comment:
although the original code does this, this is actually not correct. We should rethrow e unless `e.code ===ENOENT` otherwise if the failure is unexpected we will do the IO bellow erroneously, and hide the original error making debugging/root cause analysis way harder.


```js
try {
  fs.writeFileSync(outputPath, outputString, { encoding: outputEncoding });
} catch (e) {
  if (e !== null && typeof e === 'object' && e.code === 'ENOENT') {
   mkdirp.sync(path.dirname(outputPath));
   fs.writeFileSync(outputPath, outputString, {
     encoding: outputEncoding
   });
  } else {
    // unexpected error, simply re-throw;
    throw e;
  }
}
```